src/components/views/elements/Reply.js

```diff
@@ -0,0 +1,214 @@
+import {_t} from '../../../languageHandler';
+import PropTypes from 'prop-types';
+import dis from '../../../dispatcher';
+import MatrixClientPeg from '../../../MatrixClientPeg';
+import {wantsDateSeparator} from '../../../DateUtils';
+import {MatrixEvent} from 'matrix-js-sdk';
+import {makeUserPermalink} from "../../../matrix-to";
+import SettingsStore from "../../../settings/SettingsStore";
+
+// This component does no cycle detection, simply because the only way to make such a cycle
+// would be to craft event_id's, using a custom homeserver; even then the impact would be low
+// as each event being loaded (after the first) is triggered by an explicit user action.
+export default class Reply extends React.Component {
+    static propTypes = {
+        // The parent event
+        parentEv: PropTypes.instanceOf(MatrixEvent),
+
+        onWidgetLoad: PropTypes.func.isRequired,
+    };
+
+    constructor(props, context) {
+        super(props, context);
+
+        this.state = {
+            // The loaded events to be rendered as linear-replies
+            events: [],
+
+            // The latest loaded event which has not yet been shown
+            loadedEv: null,
+            // Whether the component is still loading more events
+            loading: true,
+
+            // Whether as error was encountered fetching a replied to event.
+            err: false,
+        };
+
+        this.onQuoteClick = this.onQuoteClick.bind(this);
+
+        this.unmounted = false;
+    }
+
+    componentWillMount() {
+        this.room = this.getRoom(this.props.parentEv.getRoomId());
+        this.initialize();
+    }
+
+    componentWillUnmount() {
+        this.unmounted = true;
+    }
+
+    async initialize() {
+        const {parentEv} = this.props;
+        const inReplyTo = Reply.getInReplyTo(parentEv);
+
+        const ev = await this.getEvent(this.room, inReplyTo['event_id']);
+        if (this.unmounted) return;
+
+        if (ev) {
+            this.setState({
+                events: [ev],
+            }, this.loadNextEvent);
+        } else {
+            this.setState({err: true});
+        }
+    }
+
+    async loadNextEvent() {
+        const ev = this.state.events[0];
+        const inReplyTo = Reply.getInReplyTo(ev);
+
+        if (!inReplyTo) {
+            if (this.unmounted) return;
+            this.setState({
+                loading: false,
+            }, this.props.onWidgetLoad);
+            return;
+        }
+
+        const loadedEv = await this.getEvent(this.room, inReplyTo['event_id']);
+        if (this.unmounted) return;
+
+        if (loadedEv) {
+            this.setState({loadedEv}, this.props.onWidgetLoad);
+        } else {
+            this.setState({err: true}, this.props.onWidgetLoad);
+        }
+    }
+
+    getRoom(id) {
+        const cli = MatrixClientPeg.get();
+        if (id[0] === '!') return cli.getRoom(id);
+
+        return cli.getRooms().find((r) => {
+            return r.getAliases().includes(id);
```

Comment:
I'm worried at the O(n^2)-ness of this. `getAliases` loops over each alias event, and once that's done we also iterate over each alias.

Reply:
Moreover, I don't see how `id` would ever be an alias so this looks like it's left over from a previous iteration.

Reply:
good catch